### ISSUE TITLE
@BeforeXXX and @AfterXXX Do Not Execute In Expected Order

### ISSUE THREAD

**Issue by staley-robert** [cbeust/testng#2330]
### TestNG Version
7.1.0

### Expected behavior
@BeforeXXX and @AfterXXX annotated methods should be executed in _correct_ order (i.e. same order they were executed in version 6.14.3).

beforeTest
beforeClass
beforeMethod
testMethod1
afterMethod
beforeMethod
testMethod2
afterMethod
beforeMethod
testMethod3
afterMethod
afterClass
afterTest

### Actual behavior
@BeforeXXX and @AfterXXX annotated methods are not executed in _correct_ order.  See sample below and attachment with working example of both 6.14.3 working correctly and 7.1.0 working incorrectly.

beforeTest
beforeClass
afterClass
afterMethod
afterTest
beforeClass
beforeMethod
beforeTest
beforeMethod
testMethod1
afterMethod
beforeMethod
testMethod2
afterMethod
beforeMethod
testMethod3
afterMethod
afterTest
afterClass
afterMethod
beforeTest
beforeClass
beforeMethod
afterClass
afterTest

### Is the issue reproductible on runner?

- [ X] Shell
- [ ] Maven
- [ ] Gradle
- [ ] Ant
- [ ] Eclipse
- [ ] IntelliJ
- [ ] NetBeans

### Test case sample
```java
@Test(groups = "group1")
public class Test1 {
    @BeforeClass
    public void beforeClass() {
        System.out.println("beforeClass");
    }

    @BeforeTest
    public void beforeTest() {
        System.out.println("beforeTest");
    }

    @BeforeMethod
    public void beforeMethod() {
        System.out.println("beforeMethod");
    }

    public void testMethod1() {
        System.out.println("testMethod1");
    }

    public void testMethod2() {
        System.out.println("testMethod2");
    }

    public void testMethod3() {
        System.out.println("testMethod3");
    }

    @AfterMethod
    public void afterMethod() {
        System.out.println("afterMethod");
    }

    @AfterTest
    public void afterTest() {
        System.out.println("afterTest");
    }

    @AfterClass
    public void afterClass() {
        System.out.println("afterClass");
    }
}
```
```xml
<!DOCTYPE suite SYSTEM "http://testng.org/testng-1.0.dtd" >
<suite name="Suite1">
    <test name="Test1">
        <groups>
            <run>
                <include name="group1"/>
            </run>
        </groups>
        <classes>
            <class name="com.test.Test1"/>
        </classes>
    </test>
</suite>
```
[testTestNgBeforeAfter.zip](https://github.com/cbeust/testng/files/4813233/testTestNgBeforeAfter.zip)

**Comment by darklynx** [comment]
this is a known issue, see:
* https://github.com/cbeust/testng/issues/2254
* https://github.com/cbeust/testng/issues/2209
* https://github.com/cbeust/testng/issues/2152
* https://github.com/cbeust/testng/issues/2156

and presumably is was fixed by:
* https://github.com/cbeust/testng/pull/2260

I'm waiting since March for a release of `ver. 7.2.0` that was promised here: https://github.com/cbeust/testng/issues/2257#issuecomment-597433537

But no updates about release so far 😢

**Comment by staley-robert** [comment]
Thanks!

**Comment by krmahadevan** [comment]
This issue is fixed in TestNG `7.3.0` which was recently released.

```java
package com.rationaleemotions.github.issue2330;

import java.lang.reflect.Method;
import org.testng.annotations.AfterClass;
import org.testng.annotations.AfterMethod;
import org.testng.annotations.AfterTest;
import org.testng.annotations.BeforeClass;
import org.testng.annotations.BeforeMethod;
import org.testng.annotations.BeforeTest;
import org.testng.annotations.Test;

@Test(groups = "group1")
public class SampleTestClass {
  @BeforeClass
  public void beforeClass() {
    System.out.println("beforeClass");
  }

  @BeforeTest
  public void beforeTest() {
    System.out.println("beforeTest");
  }

  @BeforeMethod
  public void beforeMethod(Method method) {
    System.out.println("beforeMethod ---> " + method.getName());
  }

  public void testMethod1() {
    System.out.println("testMethod1");
  }

  public void testMethod2() {
    System.out.println("testMethod2");
  }

  public void testMethod3() {
    System.out.println("testMethod3");
  }

  @AfterMethod
  public void afterMethod(Method method) {
    System.out.println("afterMethod ---> " + method.getName());
  }

  @AfterTest
  public void afterTest() {
    System.out.println("afterTest");
  }

  @AfterClass
  public void afterClass() {
    System.out.println("afterClass");
  }
}
```

```xml
<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE suite SYSTEM "https://testng.org/testng-1.0.dtd">
<suite name="2330_suite" verbose="2">
  <test name="2330_test" verbose="2">
    <groups>
      <run>
        <include name="group1"/>
      </run>
    </groups>
    <classes>
      <class name="com.rationaleemotions.github.issue2330.SampleTestClass"/>
    </classes>
  </test>
</suite>

```

Here's the execution output from `7.3.0`

```
...
... TestNG 7.3.0 by CÃ©dric Beust (cedric@beust.com)
...

beforeTest
beforeClass
beforeMethod ---> testMethod1
testMethod1
afterMethod ---> testMethod1
beforeMethod ---> testMethod2
testMethod2
afterMethod ---> testMethod2
beforeMethod ---> testMethod3
testMethod3
afterMethod ---> testMethod3
afterClass
afterTest

PASSED: testMethod1
PASSED: testMethod2
PASSED: testMethod3

===============================================
    2330_test
    Tests run: 3, Failures: 0, Skips: 0
===============================================


===============================================
2330_suite
Total tests run: 3, Passes: 3, Failures: 0, Skips: 0
===============================================

Process finished with exit code 0
```
